### DISCUSSION TITLE
Monorepo created with Turborepo and deploying it to Vercel

### DISCUSSION THREAD

**Discussion by MiguelAngelGarciaGomez** [vercel/community#4288] [Help] [answered]
### Summary

This is more about a potential approach to creating a monorepo using Turborepo, and subsequently deploying it with Vercel.

Suppose you have three distinct Next.js projects, each utilizing similar libraries and packages. My proposition is to consolidate these projects into a monorepo using Turborepo, and house all commonly shared resources (like services and other similar elements) within a designated `shared` folder.

This would result in a monorepo with a structure resembling the following:

```
project
|- apps
    |- app1
        |- package.json
    |- app2
        |- package.json
    |- app3
         |- package.json
|- packages
    |- shared-code
         |- package.json
|- package.json
```

The `package.json` for every `appX` is something like:
```
{
  "name": "app1",
  "version": "1.0.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  }
}
``` 
And the `package.json` for the root project is something like:
```
{
  "name": "root-app",
  "version": "1.0.0",
  "private": true,
  "workspaces": [
    "apps/*"
  ],
  "scripts": {
    "build": "turbo run build",
    "dev": "dotenv -- turbo run dev",
    "start": "turbo run start",
    "lint": "turbo run lint",
    "clean": "turbo run clean"
  },
  "dependencies": {
    "@types/jsonwebtoken": "^9.0.3",
    "@types/node": "20.6.3",
    "@types/react": "18.2.22",
    "@types/react-dom": "18.2.7",
    "dotenv-cli": "^7.3.0",
    "eslint": "8.49.0",
    "eslint-config-next": "13.5.2",
    [...]
    "next": "13.5.2",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "swr": "^2.2.2",
    "typescript": "5.2.2"
  },
  "devDependencies": {
    "@types/lodash-es": "^4.17.9",
    "encoding": "^0.1.13",
    "eslint-config-turbo": "latest",
    "turbo": "^1.10.14"
  }
}
```
Here's where my first question arises. I've situated all the shared libraries (Next.js included) within the root `package.json` file, as all applications (`app1`, `app2`, and `app3`) are intended to leverage the same libraries and Next.js version. Is this the correct methodology? Or should each individual application define its own Next.js libraries within their respective `package.json` files?

This inquiry is largely prompted by a hurdle I've encountered when attempting to deploy the structure on Vercel, which results in the following error message:
```
13:26:49.870 | Running "vercel build"
13:26:50.365 | Vercel CLI 32.3.1
13:26:50.501 | > Detected Turbo. Adjusting default settings...
13:26:50.985 | Running "install" command: `npm install --prefix=`...
13:26:52.388 |  
13:26:52.388 | added 17 packages, and audited 18 packages in 1s
13:26:52.388 |  
13:26:52.388 | 2 packages are looking for funding
13:26:52.388 | run `npm fund` for details
13:26:52.389 |  
13:26:52.389 | found 0 vulnerabilities
13:26:52.399 | Warning: Could not identify Next.js version, ensure it is defined as a project dependency.
13:26:53.553 | Error: No Next.js version could be detected in your project. Make sure `"next"` is installed in "dependencies" or "devDependencies"
```

Maybe this issue: https://github.com/orgs/vercel/discussions/4190 is related to this.

### Steps to Reproduce

* Create 3 Next.js projects.
* Create a monorepo with Turborepo.
* Copy the 3 created projects in the `apps` folder of the monorepo created.
* Delete all the dependencies in the `package.json` of every Next.js project created.
* Copy all these dependencies in the `package.json` of the root project.
* Create a new Vercel project.
* Connect the Vercel project with the monorepo project.
* While configuring the project, select "Root directory" as one of the apps (`apps/app1` for example").
* Click on "Deploy" and the message:
```
13:26:52.399 | Warning: Could not identify Next.js version, ensure it is defined as a project dependency.
13:26:53.553 | Error: No Next.js version could be detected in your project. Make sure `"next"` is installed in "dependencies" or "devDependencies"
```
Should appear.

**Comment by amyegan** [comment]
Hi @MiguelAngelGarciaGomez! Have you seen the documentation on [using monorepos with Vercel](https://vercel.com/docs/monorepos) and the [Turborepo examples](https://vercel.com/templates?type=monorepos)? Those should help you get your repo set up and projects deployed.

**Reply by MiguelAngelGarciaGomez** [reply]
Hi @amyegan!

Thanks a lot for the links! I have followed the documentation but I haven't checked the Turborepo examples. Having checked the [Turborepo & Next.js Starter repo](https://github.com/vercel/turbo/tree/main/examples/basic) I have seen that the two Next.js apps have their own `package.json` with all the Next.js and React dependencies and [the root `package.json`](https://github.com/vercel/turbo/blob/main/examples/basic/package.json) only have some `devDependencies`. 
Therefore, I think that my solution is wrong and the three apps should have their own `package.json` with Next.js and React dependencies, right?
And just out curiosity, what I posted in my first message, could it be possible? I mean, to have only one "centralised" `package.json` file so all the sub-projects have the same dependencies.
Thanks again!

**Reply by amyegan** [reply]
That's more related to configuring [npm workspaces](https://docs.npmjs.com/cli/v10/configuring-npm/package-json#workspaces). The Turborepo docs have more info about that: https://turbo.build/repo/docs/handbook/workspaces

**Reply by MiguelAngelGarciaGomez** [reply]
Many thanks @amyegan again for the documentation ;) 

Though I've read your response, I must apologise as I still harbour some uncertainties regarding the approach I previously proposed.

Let me rephrase my question: If the Next.js applications within the apps directory all utilise the same Next.js version, would it be best to define this within the root `package.json`, especially if each component is intended to be independently deployed on Vercel? Alternatively, should each Next.js application specify its own Next.js version within its respective `package.json` file? My concern is that the latter approach might compromise maintainability, considering an upgrade of Next.js would necessitate changes across all three `package.json` files.

Thank you!

**Reply by mehulkar** [reply, marked as answer]
@MiguelAngelGarciaGomez you should put each shared dependency in each app's `package.json`. Turborepo doesn't support (aka doesn't work well) if you put common dependencies in your root package.json.
